[VectorCombine] Add constant splat handling for shuffleToIdentity

davemgreen opened the pull request:
This just adds splat constants, which can be treated like any other splat which hopefully makes them very simple. It does not try to handle more complex constant vectors yet, just the more common splats.

I will rebase over #92766, but already had the patch for this version.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1739,6 +1739,15 @@ bool VectorCombine::foldShuffleToIdentity(Instruction &I) {
       continue;
     }
+    // Look for constants, for the moment only supporting constant splats.
+    if (isa<Constant>(Item[0].first) &&
+        cast<Constant>(Item[0].first)->getSplatValue() &&
```
```suggestion
    if (auto *C = dyn_cast<Constant>(Item[0].first); C &&
        C->getSplatValue() &&
```

davemgreen committed:
[VectorCombine] Add constant splat handling for shuffleToIdentity

This adds splat constants, which can be treated like any other splat which
makes them very simple. Other constants are not handled yet, they are planned
in a later patch.

artagnon reviewed (commented):
Apart from renaming variables, which would be a nice-to-fix, this LGTM.

RKSimon reviewed (approved):
LGTM
